pf: Override description list; condense & subgridify

garrett opened the pull request:
1. Allow actions to wrap in constrained space (at the card level)
2. Fix too-much-space between definition list items
3. Use subgrid to fix alignment of definition list labels to the longest one

garrett commented:
The subgrid only currently works in Firefox. The layout still needs additional fixes for non-Firefox browsers (Chromium and WebKit engines), but I wanted to share the fixes I already have before the long weekend.

The Chromium/WebKit fixes can land in this PR or in an additional follow-up.

garrett commented:
This PR fully addresses the issues I brought up in https://github.com/cockpit-project/cockpit/pull/14346#issuecomment-709206301 for Firefox. It addresses a lot of them in other browsers too, but needs additional work (perhaps by setting a breakpoint value on the flex wrapping the description lists).

KKoukiou commented on pkg/machines/components/vm/vmExpandedContent.scss:
```diff
@@ -55,6 +55,30 @@
     }
   }
 
+  .overview-tab .pf-c-description-list {
```
I don't understand why this is needed here. 
Description list (by itself) works correctly by wrapping when in small screens.

garrett replied:
The labels are exceedingly long and shove the content to be way too wide, causing content to be shoved way offscreen.

The problem is that the description list is not by itself; it is side-by-side with another, and since they're both much larger (taller and, more importantly, wider) than when we worked on the PR originally, it breaks the page.

We still need to fix it for non-Firefox browsers (everything that does not _yet_ support subgrid), of course.

KKoukiou replied:
Do you think it makes sense to have a global override (patternfly-4-overrides)  for the description list making it less wide then?

I have similar issue in other PR, https://github.com/cockpit-project/cockpit/pull/14754/commits/f0027071cb50b206c46d31b34df56481500a3393#diff-325ba8d07b7191456e22e2785c61be383372356c4e38b9a7cbfe35021be2af96R50

And it would be nice to keep it globally consistent.

garrett replied:
Yeah, I think so. 

We should try to upstream both, and meanwhile, apply (probably) both at a Cockpit-global override.

The only problem I really see is that — for the time being — only Firefox would have the optimized subgrid layout, so there would be two codepaths. It's not _that_ different, but it is something to consider. Hopefully Chromium and WebKit will _finally_ get subgrid in the next few months. (They're both supposed to.)

https://blog.chromium.org/2020/06/improving-chromiums-browser.html

> Chromium still doesn't support subgrid.
>
> Hopefully it won't be an exception for much longer. It's still early days, but I'm excited to share that a team at Microsoft Edge are working on rearchitecting Chromium's Grid support to use the new LayoutNG engine - and as part of this are intending to add subgrid support! We're excited to see this feature land and want to express our appreciation to our partners at Microsoft, Igalia, and the many other Chromium contributors who have worked on CSS Grid support.

And it's on WebKit's radar: https://bugs.webkit.org/show_bug.cgi?id=202115

KKoukiou commented on pkg/machines/components/vmOverviewTabLibvirt.jsx:
```diff
@@ -269,9 +269,9 @@ class VmOverviewTabLibvirt extends React.Component {
                             <DescriptionListGroup>
                                 <DescriptionListTerm>{_("State")}</DescriptionListTerm>
-                                <DescriptionListGroup>
+                                <DescriptionListDescription>
```
good catch.

KKoukiou replied:
we cherry-picked it and it's in master already

marusak commented:
> I still have some work to do to finish it up

Changing to draft then

garrett commented:
I think we do want these fixes to land before the release; they're an improvement in all browsers already.

If we could move some of them to be more general, it would be awesome, but we could do that in a follow-up to this.

KKoukiou commented:
> I think we do want these fixes to land before the release; they're an improvement in all browsers already.
> 
> If we could move some of them to be more general, it would be awesome, but we could do that in a follow-up to this.

I already cherry-picked the commit that fixes the DescriptionListGroup wrong usage and it's merged.

garrett commented:
Do we have a fix for the extra spacing in master too?

I think the thing I wanted to do for Chrome & WebKit was find a better size for the width; basically to prevent the items from going way too wide and messing up the layout.

KKoukiou commented:
> > I think we do want these fixes to land before the release; they're an improvement in all browsers already.
> > If we could move some of them to be more general, it would be awesome, but we could do that in a follow-up to this.
> 
> I already cherry-picked the commit that fixes the DescriptionListGroup wrong usage and it's merged.

If you want changes in the description list component, these should go in the pkg/lib/patternfly/patternfly-4-overrides.scss and not in the machines specific files.

KKoukiou commented:
> Do we have a fix for the extra spacing in master too?
> 
> I think the thing I wanted to do for Chrome & WebKit was find a better size for the width; basically to prevent the items from going way too wide and messing up the layout.

We do have a width reduce since last week, see https://github.com/cockpit-project/cockpit/blob/master/pkg/lib/patternfly/patternfly-4-overrides.scss#L52

garrett commented:
> these should go in the pkg/lib/patternfly/patternfly-4-overrides.scss and not in the machines specific files

Right, which is why this PR needs some additional work.

> We do have a width reduce since last week, see https://github.com/cockpit-project/cockpit/blob/master/pkg/lib/patternfly/patternfly-4-overrides.scss#L52

Ah, nice!

garrett commented:
Rebased and moved description list to PF4 override. (Still need to remove traces of it from Machines in an earlier commit.)

Still left to do: 
- Better Chrome/WebKit layout (might be done locally in Machines)

KKoukiou commented:
Is this still valid PR?

garrett commented:
I think it's still valid. I just rebased on master and squashed down the subgrid override. I need to check to see what Chrome and WebKit do, however.

This could be split into two PRs, I suppose: One for the card wrapping (which fixes smaller screens with larger content) and the PF override to use subgrids to actually align, instead of hardcoding fixes.

garrett commented:
Moved the card header wrap override to https://github.com/cockpit-project/cockpit/pull/14988 and opened upstream PF issue and PR with a simple fix.

garrett commented:
Note: Most of this PR was moved to other PRs; what remains is adding subgrid to pf-c-description-list.

KKoukiou commented on pkg/lib/patternfly/patternfly-4-overrides.scss:
```diff
@@ -39,11 +39,31 @@ svg {
-    // The default gap between the rows is too large
-    --pf-c-description-list--RowGap: 1rem;
+    // Decrease the vertical spacing between lines
+    --pf-c-description-list--RowGap: var(--pf-global--spacer--sm);
```
To me it seems this is too little spacing, especially because the space between the label and the value remains big.
Before:

![Screen Shot 2021-02-04 at 10 53 32](https://user-images.githubusercontent.com/14921356
/106875909-79a83280-66d7-11eb-92c9-aca2b06f0daa.png)

After:

![Screen Shot 2021-02-04 at 10 52 57](https://user-images.githubusercontent.com/14921356/106875912-7a40c900-66d7-11eb-964a-1de6f236c9b5.png)

And vertical lists also are too condensed I feel:

Before:

![Screen Shot 2021-02-04 at 10 57 47](https://user-images.githubusercontent.com/14921356/106876310-dd326000-66d7-11eb-9f92-eb274ef3e232.png)

After: 

![Screen Shot 2021-02-04 at 10 57 43](https://user-images.githubusercontent.com/14921356/106876314-ddcaf680-66d7-11eb-948a-5bde76314ca4.png)

garrett replied:
Yeah, I think this PR predates the vertical lists? (Or we just weren't using them?) Anyway, it was made with horizontal ones in mind.

KKoukiou reviewed (commented):
I am hesitant about the 'condence' part

martinpitt commented:
Moving to draft for now - it seems this needs some more discussion/iteration?

garrett commented:
Yep, that makes sense. Everything non-controversial was moved out to separate PRs.

garrett commented:
I'll just give up on this, just like with https://github.com/cockpit-project/cockpit/issues/12843#issuecomment-779341505. Browser uptake of subgrid is just too slow (nonexistant outside of Firefox).
